#161342852 Add Search Functionality and Custom filtering

ucheg6 opened the pull request:
#### What does this PR do?
- Add Search Functionality and Custom filtering based on parameters
#### Description of Task to be completed?
- create a migration file to alter articles tables with a column searchVectors that contains ts_vectors dataType
- create an index and a trigger in the articles table to collate info from title, body and description
- create a function in the article controller that takes in searchTerm, author or tag parameters
- filter search based on the availability of the above parameters
#### How should this be manually tested?
- Git clone the repository
- cd into the `haven-ah-backend` folder
- run `npm run start-dev`
- Fire up the route `localhost:5000/api/v1/articles/search?:queries` as a post request
- To test the entire application, run `npm run test`
#### Any background context you want to provide?
N/A
#### What are the relevant pivotal tracker stories?
[ #161342852](https://www.pivotaltracker.com/story/show/161342852)
#### Screenshots (if appropriate)
<img width="1119" alt="screen shot 2018-11-14 at 1 58 28 pm" src="https://user-images.githubusercontent.com/34472497/48484291-0bd17100-e816-11e8-9b40-244ff1e5d28c.png">
<img width="1117" alt="screen shot 2018-11-14 at 1 59 06 pm" src="https://user-images.githubusercontent.
<img width="1116" alt="screen shot 2018-11-14 at 2 00 04 pm" src="https://user-images.githubusercontent.com/34472497/48484389-4a672b80-e816-11e8-8931-59ba3838f37d.png">
<img width="1095" alt="screen shot 2018-11-14 at 2 00 31 pm" src="https://user-images.githubusercontent.com/34472497/48484402-5226d000-e816-11e8-9313-b45f85f4387d.png">

com/34472497/48484304-13911580-e816-11e8-914d-27331394dd22.png">
<img width="1100" alt="screen shot 2018-11-14 at 1 59 25 pm" src="https://user-images.githubusercontent.com/34472497/48484357-315e7a80-e816-11e8-84bd-a70d0f9178a2.png">
<img width="1128" alt="screen shot 2018-11-14 at 1 59 42 pm" src="https://user-images.githubusercontent.com/34472497/48484376-3f140000-e816-11e8-8db2-9483121e9405.png">

#### Questions:
N/A

victor-nweze-deel reviewed (commented):
Nice work @ucheg6

TheoOkafor commented on server/repository/articleRepository.js:
```diff
@@ -110,6 +112,86 @@ class ArticleRepository {
+
+    return keywordsResults;
+  }
 }
```
Nice work here @ucheg6 👍

TheoOkafor reviewed (commented):
Great job @ucheg6

seyi-adeleke commented on server/controllers/article.js:
```diff
@@ -150,6 +150,25 @@ class Article {
+   */
+  static async searchArticle(request, response) {
+    const { author, tag, keywords } = request.query;
+    const foundArticle = await articleRepo.search(keywords, author, tag);
```
are you returning only one article?

ucheg6 replied:
Nope, I'll handle this. thanks

seyi-adeleke commented on server/controllers/article.js:
```diff
@@ -150,6 +150,25 @@ class Article {
+  static async searchArticle(request, response) {
+    const { author, tag, keywords } = request.query;
+    const foundArticle = await articleRepo.search(keywords, author, tag);
+    if (foundArticle.length === 0) {
```
`!foundArticle.length  `

ucheg6 replied:
This worked Thanks.

seyi-adeleke commented on server/middlewares/validations.js:
```diff
@@ -235,6 +235,29 @@ class inputValidator {
+   * @param {object} next callback function
+   * @returns {object} http response object
+   */
+  static searchValidator(request, response, next) {
```
shouldnt you validate that author, tag, and keywords are strings?

ucheg6 replied:
This has been done

ucheg6 committed:
feat(search): Add Search Functionality and Custom filtering

- create a migration file to alter articles tables with a column searchVectors that contains ts_vectors dataType
- create an index and a trigger in the articles table to collate info from title, body and description
- create a function in the article controller that takes in searchterm, author or tag parameters
- filter search based on availability of the above parameters
